torch/autograd/_functions/stochastic.py

```diff
@@ -30,9 +30,10 @@ def backward(self, reward):
         output_probs = probs.gather(1, samples)
         output_probs.add_(1e-6).reciprocal_()
         output_probs.neg_().mul_(reward)
-        # TODO: add batched index_add
-        for i in range(probs.size(0)):
-            grad_probs[i].index_add_(0, samples[i], output_probs[i])
+        # Fill in gradients
+        idx = grad_probs.new().resize_(grad_probs.size(0)).fill_(1)
```

Comment:
a more efficient way to write this code is:

```python
idx = grad_probs.new().long().resize_(grad_probs.size(0))
torch.arange(0, grad_probs.size(0), out=idx)
grad_probs[idx, samples.t()] += output_probs.t()
```

However, your change is not correct. I think advanced indexing doesn't handle collissions well (two adds on same element): cc @killeent do you know if we handle collissions?

Here's a simple test I wrote to verify that the old and new results differ:

```python
import torch

probs = torch.randn(2, 5).add_(10)
samples = probs.multinomial(10, True)

grad_probs = probs.new().resize_as_(probs).zero_()
grad_probs.normal_() # some random gradients

output_probs = probs.gather(1, samples)
output_probs.add_(1e-6).reciprocal_()

# now save original grad_probs
grad_probs_old = grad_probs.clone()


# do it in the method of existing code
for i in range(probs.size(0)):
    grad_probs[i].index_add_(0, samples[i], output_probs[i])

grad_probs_unbatched = grad_probs.clone()
print(grad_probs_unbatched)

# restore grad_probs again, for batched method
grad_probs = grad_probs_old.clone()
idx = grad_probs.new().resize_(grad_probs.size(0)).fill_(1)
idx = idx.cumsum(0) - 1
grad_probs[idx.long(), samples.t()] += output_probs.t()


print(grad_probs)
print(grad_probs_manual)
```

Reply:
No, the advanced indexing only takes the last element to be assigned, and do not sum them.
I thought about this during the advanced indexing PR, but I decided to not comment on that as pytorch has the same behaviour as numpy.